tests/integration/test_block.py

```diff
@@ -71,7 +42,226 @@ def _check(a, b, depth, sizes):
+            [zero_2d],
+        ]
+
+        print_msg = "np & cunumeric.block()"
+        check_array_method("block", [arg], {}, print_msg)
+
+    def test_nested(self):
+        one = np.array([1, 1, 1])
+        two = np.array([[2, 2, 2], [2, 2, 2], [2, 2, 2]])
+        three = np.array([3, 3, 3])
+        four = np.array([4, 4, 4])
+        five = np.array([5])
+        six = np.array([6, 6, 6, 6, 6])
+        zero = np.zeros((2, 6))
+
+        result = num.block(
+            [[num.block([[one], [three], [four]]), two], [five, six], [zero]]
+        )
+        expected = np.array(
+            [
+                [1, 1, 1, 2, 2, 2],
+                [3, 3, 3, 2, 2, 2],
+                [4, 4, 4, 2, 2, 2],
+                [5, 6, 6, 6, 6, 6],
+                [0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0],
+            ]
+        )
+
+        assert np.array_equal(result, expected)
+
+    def test_3d(self):
+        a000 = np.ones((2, 2, 2), int) * 1
+
+        a100 = np.ones((3, 2, 2), int) * 2
+        a010 = np.ones((2, 3, 2), int) * 3
+        a001 = np.ones((2, 2, 3), int) * 4
+
+        a011 = np.ones((2, 3, 3), int) * 5
+        a101 = np.ones((3, 2, 3), int) * 6
+        a110 = np.ones((3, 3, 2), int) * 7
+
+        a111 = np.ones((3, 3, 3), int) * 8
+
+        arg = [
+            [
+                [a000, a001],
+                [a010, a011],
+            ],
+            [
+                [a100, a101],
+                [a110, a111],
+            ],
+        ]
+
+        print_msg = "np & cunumeric.block()"
+        check_array_method("block", [arg], {}, print_msg, check_type=False)
+
+
+class TestBlockErrors:
+    def test_mismatched_shape_1(self):
+        msg = "All arguments to block must have the same number of dimensions"
+        a = np.array([0, 0])
+        b = np.eye(2)
+        with pytest.raises(ValueError, match=msg):
+            num.block([a, b])
+        with pytest.raises(ValueError, match=msg):
+            num.block([b, a])
+
+    def test_mismatched_shape_2(self):
+        a = np.array([[0, 0]])
+        b = np.eye(2)
+        with pytest.raises(ValueError):
+            num.block([a, b])
+
+    @pytest.mark.xfail
+    def test_mismatched_shape_3(self):
+        a = np.array([[0, 0]])
+        b = np.eye(2)
+
+        # numpy: raises ValueError
+        # cumunerics: pass, output is [[1, 0, 0, 0]
+        #                              [0, 1, 0, 0]]
+        with pytest.raises(ValueError):
+            num.block([b, a])
+
+    def test_no_lists(self):
+        # numpy: pass, output is np.array(1)
+        # cunumeric: raises TypeError, cunumeric doesn't support 0-D array
+        # assert np.array_equal(num.block(1), np.array(1))
+
+        # numpy: pass, output is np.eye(3)
+        # cunumeric: pass, output is 1-D array: [1, 0, 0, 0, 1, 0, 0, 0, 1]
+        # assert np.array_equal(num.block(np.eye(3)), np.eye(3))
+        np.array_equal(num.block(num.eye(3)), [1, 0, 0, 0, 1, 0, 0, 0, 1])
```

Comment:
At least the first bullet would be great I think; it will also notify us if the behavior ever changes. And we can keep track of divergences, in case we want to fix them.

Taking a quick look at these divergences specifically, it looks like we may want to fix them. @robinw0928 can you open issues for them?

Reply:
I've added a link to this comment thread there, I don't  want to hold up this PR, we can add the `xfail` tests as a dedicated chunk if work in another PR